web/ui_utils.js

```diff
@@ -153,7 +153,8 @@ function scrollIntoView(element, spot, skipOverflowHiddenElements = false) {
   }
   let offsetY = element.offsetTop + element.clientTop;
   let offsetX = element.offsetLeft + element.clientLeft;
-  while (parent.clientHeight === parent.scrollHeight ||
+  while ((parent.clientHeight === parent.scrollHeight &&
+         parent.clientWidth === parent.scrollWidth) ||
```

Comment:
This `while` loop is walking the ancestry of the provided element, looking for the first ancestor that has scroll bars, so that said ancestor can have its `scrollTop` and (possibly) `scrollLeft` adjusted so that the provided element is in view. Prior to this change, the `parent.clientHeight === parent.scrollHeight` condition only checks for the existence of vertical scroll bars. If in a situation where the element is contained in a viewport with only a horizontal scroll bar, this function would return without scrolling the element into view. Prior to implementing this feature, it was very unlikely (but possible!) to be viewing a PDF with a horizontal scroll bar but no vertical scroll bar, so this bug would probably have gone unnoticed (to reproduce it on the master branch, you'd need to have a PDF that is wider than the viewport but not taller, *and* you'd need to be clicking a link or something that should cause a scroll to a specific point on the page), but that's why this change is implemented unconditionally—IMO it's technically a bug even in the original vertical scrolling mode, just one that is exacerbated by the other modes implemented in this PR.

Reply:
Thanks for this explanation. You're right, I have created a test PDF that you described and can observe that there is no scrolling into view with the current master (buggy) when the vertical scrollnbar is absent and the horizontal scrollbar is present.

Could you amend the commit message to also state that you've fixed this bug?

To test, I did the following:

1. Create PDF of 90cm wide and 10cm tall, with a link at the left and right (both pointing to a footnote at the bottom left corner):

    ```
    echo '\documentclass{article}\usepackage[paperwidth=90cm,paperheight=10cm]{geometry}\usepackage{hyperref}\begin{document}\footnote{left}\hfill\footnote{right}\end{document}' > page.tex && pdflatex page.tex
    ```
    Result: [page.pdf](https://github.com/mozilla/pdf.js/files/1729606/page.pdf)

2. Open the PDF in the viewer (master branch & your fork).